spec/models/time_entry_spec.rb

```diff
@@ -351,5 +351,73 @@
         it { expect(TimeEntry.visible(time_entry2.user, project).all).to match_array([time_entry2]) }
       end
     end
+
+    context 'calculate dates' do
+      let(:my_role) { FactoryGirl.create(:role, permissions: [:view_own_time_entries]) }
+      let(:user3) do
+        FactoryGirl.create(:user, member_in_project: project, member_through_role: my_role)
+      end
+
+      let(:late_time_entry) do
+        FactoryGirl.create(:time_entry,
+                           project: project,
+                           work_package: work_package,
+                           spent_on: 2.days.ago,
+                           hours: hours,
+                           user: user3,
+                           rate: hourly_one,
+                           comments: 'ipsum')
+      end
+      let(:early_time_entry) do
+        FactoryGirl.create(:time_entry,
+                           project: project,
+                           work_package: work_package,
+                           spent_on: 4.days.ago,
+                           hours: hours,
+                           user: user3,
+                           rate: hourly_one,
+                           comments: 'dolor')
+      end
+      let(:other_time_entry) do
+        FactoryGirl.create(:time_entry,
+                           project: project,
+                           work_package: work_package,
+                           spent_on: 6.days.ago,
+                           hours: hours,
+                           user: user2,
+                           rate: hourly_one,
+                           comments: 'dolor')
+      end
+      let(:another_time_entry) do
+        FactoryGirl.create(:time_entry,
+                           project: project,
+                           work_package: work_package,
+                           spent_on: 1.days.ago,
+                           hours: hours,
+                           user: user2,
+                           rate: hourly_one,
+                           comments: 'dolor')
+      end
+
+      before do
+        early_time_entry.save!
+        late_time_entry.save!
+        other_time_entry.save!
+        another_time_entry.save!
+
+        allow(User).to receive(:current).and_return(user3)
+      end
+      describe '#earliest_date_for_project' do
```

Comment:
Prefer double-quoted strings unless you need single quotes to avoid extra backslashes for escaping.